### ISSUE TITLE
[QUESTION] (Who is to blame?) Winit, Vulkano and/or me?

### ISSUE THREAD

**Issue by DancingBard** [tomaka/winit#582]
First off, I am sorry if I missed something obvious. It's just that during the upgrade to `winit 0.16` something strange happened within my pet-project and I boiled it down to the following mini example.

I do not understand what's happening, probably everything is alright and I am the one to blame. So I am here to ask and learn!

That being said, right onto the cheese:

**(And don't worry. The actual relevant code is just a few short lines. I just wanted this to be self-contained!)**

Cargo.toml:

	cargo-features = ["edition"]

	[package]
	edition = '2018'

	name = "test"
	version = "0.0.1"

	[[bin]]
	name = "test"
	path = "src/main.rs"

	[dependencies]
	vulkano = "^0.9"
	vulkano-shader-derive = "^0.9"
	winit = "^0.16"

main.rs:

	#![feature(rust_2018_preview, use_extern_macros, nll)]

	use vulkano;
	use vulkano::{
	    device::{
	        Device,
	        Queue,
	    },
	    framebuffer::RenderPassAbstract,
	    image::SwapchainImage,
	    instance::{
	        Instance,
	        PhysicalDevice,
	    },
	    swapchain::{
	        PresentMode,
	        Surface,
	        SurfaceTransform,
	        Swapchain,
	        SwapchainCreationError,
	    },
	    single_pass_renderpass,
	    ordered_passes_renderpass,
	};

	use winit;
	use winit::EventsLoop;
	use std::sync::Arc;

	use self::vulkano_win::VkSurfaceBuild;

	fn main() {
	    let window_width: u32 = 1920;
	    let window_height: u32 = 1200;

	    let size = winit::dpi::LogicalSize::new(window_width as f64, window_height as f64);
	    let (width, height): (u32, u32) = size.into();
	    println!("width: {}, height: {}", width, height);

	    let instance: Arc<Instance> = {
	        let extensions = vulkano_win::required_extensions();
	        Instance::new(None, &extensions, None).expect("failed to create Vulkan instance")
	    };

	    let cloned_instance = instance.clone();

	    let physical: PhysicalDevice = vulkano::instance::PhysicalDevice::enumerate(&cloned_instance)
	        .next().expect("no device available");

	    let events_loop = winit::EventsLoop::new();
	    let surface = winit::WindowBuilder::new()
	        .with_dimensions(size.clone())
	        .with_title("Test")
	        .build_vk_surface(&events_loop, instance.clone()).unwrap();

	    let caps = surface.capabilities(physical)
	        .expect("failed to get surface capabilities");
	    let surface_dimensions = caps.current_extent.unwrap_or([width, height]);
	    let inner_size = surface.window().get_inner_size();
	    let outer_size = surface.window().get_outer_size();
	    println!("width: {}, height: {}", width, height);
	    println!("surface_dimensions: {:?}", surface_dimensions);
	    println!("inner size: {:?}", inner_size);
	    println!("outer_size: {:?}", outer_size);
	}

	mod vulkano_win {
	    // Taken from https://github.com/vulkano-rs/vulkano/blob/master/vulkano-win/src/lib.rs in order to use up2date winit together with vulkano
	    use winit;
	    use std::error;
	    use std::fmt;
	    #[cfg(target_os = "windows")]
	    use std::ptr;
	    use std::sync::Arc;

	    use vulkano::instance::Instance;
	    use vulkano::instance::InstanceExtensions;
	    use vulkano::swapchain::Surface;
	    use vulkano::swapchain::SurfaceCreationError;
	    use winit::{EventsLoop, WindowBuilder};
	    use winit::CreationError as WindowCreationError;

	    #[cfg(target_os = "macos")]
	    use cocoa::appkit::{NSView, NSWindow};
	    #[cfg(target_os = "macos")]
	    use cocoa::base::id as cocoa_id;
	    #[cfg(target_os = "macos")]
	    use metal::CoreAnimationLayer;
	    #[cfg(target_os = "macos")]
	    use objc::runtime::YES;

	    #[cfg(target_os = "macos")]
	    use std::mem;

	    pub fn required_extensions() -> InstanceExtensions {
	        let ideal = InstanceExtensions {
	            khr_surface: true,
	            khr_xlib_surface: true,
	            khr_xcb_surface: true,
	            khr_wayland_surface: true,
	            khr_mir_surface: true,
	            khr_android_surface: true,
	            khr_win32_surface: true,
	            mvk_ios_surface: true,
	            mvk_macos_surface: true,
	            ..InstanceExtensions::none()
	        };

	        match InstanceExtensions::supported_by_core() {
	            Ok(supported) => supported.intersection(&ideal),
	            Err(_) => InstanceExtensions::none(),
	        }
	    }

	    pub trait VkSurfaceBuild {
	        fn build_vk_surface(self, events_loop: &EventsLoop, instance: Arc<Instance>)
	                            -> Result<Arc<Surface<winit::Window>>, CreationError>;
	    }

	    impl VkSurfaceBuild for WindowBuilder {
	        fn build_vk_surface(self, events_loop: &EventsLoop, instance: Arc<Instance>)
	                            -> Result<Arc<Surface<winit::Window>>, CreationError> {
	            let window = self.build(events_loop)?;
	            Ok(unsafe { winit_to_surface(instance, window) }?)
	        }
	    }

	    /// Error that can happen when creating a window.
	    #[derive(Debug)]
	    pub enum CreationError {
	        /// Error when creating the surface.
	        SurfaceCreationError(SurfaceCreationError),
	        /// Error when creating the window.
	        WindowCreationError(WindowCreationError),
	    }

	    impl error::Error for CreationError {
	        #[inline]
	        fn description(&self) -> &str {
	            match *self {
	                CreationError::SurfaceCreationError(_) => "error while creating the surface",
	                CreationError::WindowCreationError(_) => "error while creating the window",
	            }
	        }

	        #[inline]
	        fn cause(&self) -> Option<&error::Error> {
	            match *self {
	                CreationError::SurfaceCreationError(ref err) => Some(err),
	                CreationError::WindowCreationError(ref err) => Some(err),
	            }
	        }
	    }

	    impl fmt::Display for CreationError {
	        #[inline]
	        fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
	            write!(fmt, "{}", error::Error::description(self))
	        }
	    }

	    impl From<SurfaceCreationError> for CreationError {
	        #[inline]
	        fn from(err: SurfaceCreationError) -> CreationError {
	            CreationError::SurfaceCreationError(err)
	        }
	    }

	    impl From<WindowCreationError> for CreationError {
	        #[inline]
	        fn from(err: WindowCreationError) -> CreationError {
	            CreationError::WindowCreationError(err)
	        }
	    }

	    #[cfg(target_os = "android")]
	    unsafe fn winit_to_surface(instance: Arc<Instance>, win: winit::Window)
	                               -> Result<Arc<Surface<winit::Window>>, SurfaceCreationError> {
	        use winit::os::android::WindowExt;
	        Surface::from_anativewindow(instance, win.get_native_window(), win)
	    }

	    #[cfg(all(unix, not(target_os = "android"), not(target_os = "macos")))]
	    unsafe fn winit_to_surface(instance: Arc<Instance>, win: winit::Window)
	                               -> Result<Arc<Surface<winit::Window>>, SurfaceCreationError> {
	        use winit::os::unix::WindowExt;
	        match (win.get_wayland_display(), win.get_wayland_surface()) {
	            (Some(display), Some(surface)) => Surface::from_wayland(instance,
	                                                                    display,
	                                                                    surface,
	                                                                    win),
	            _ => {
	                // No wayland display found, check if we can use xlib.
	                // If not, we use xcb.
	                if instance.loaded_extensions().khr_xlib_surface {
	                    Surface::from_xlib(instance,
	                                       win.get_xlib_display().unwrap(),
	                                       win.get_xlib_window().unwrap() as _,
	                                       win)
	                } else {
	                    Surface::from_xcb(instance,
	                                      win.get_xcb_connection().unwrap(),
	                                      win.get_xlib_window().unwrap() as _,
	                                      win)
	                }
	            },
	        }
	    }

	    #[cfg(target_os = "windows")]
	    unsafe fn winit_to_surface(instance: Arc<Instance>, win: winit::Window)
	                               -> Result<Arc<Surface<winit::Window>>, SurfaceCreationError> {
	        use winit::os::windows::WindowExt;
	        Surface::from_hwnd(instance,
	                           ptr::null() as *const (), // FIXME
	                           win.get_hwnd(),
	                           win)
	    }

	    #[cfg(target_os = "macos")]
	    unsafe fn winit_to_surface(instance: Arc<Instance>, win: winit::Window)
	                               -> Result<Arc<Surface<winit::Window>>, SurfaceCreationError> {
	        use winit::os::macos::WindowExt;

	        let wnd: cocoa_id = mem::transmute(win.get_nswindow());

	        let layer = CoreAnimationLayer::new();

	        layer.set_edge_antialiasing_mask(0);
	        layer.set_presents_with_transaction(false);
	        layer.remove_all_animations();

	        let view = wnd.contentView();

	        layer.set_contents_scale(view.backingScaleFactor());
	        view.setLayer(mem::transmute(layer.as_ref())); // Bombs here with out of memory
	        view.setWantsLayer(YES);

	        Surface::from_macos_moltenvk(instance, win.get_nsview() as *const (), win)
	    }
	}

The output is: 
width: 1920, height: 1200
width: 1920, height: 1200
surface_dimensions: [1920, 1200]
inner size: Some(LogicalSize { width: 1152.0, height: 720.0 })
outer_size: Some(LogicalSize { width: 1156.8, height: 739.8 })

My expectation would be that the inner_size (or at least the outer_size) is 1920 x 1200 too.
Why isn't that the case? What am I missing here?

**Comment by francesca64** [comment]
Would the return of `get_hidpi_factor` happen to be 1.6666666666666667?

**Comment by francesca64** [comment]
Also, which platforms does this happen on?

**Comment by DancingBard** [comment]
Hey @francesca64, thanks for taking the time to investigate:

1) Yes, `surface.window().get_hidpi_factor()` returns `1.6666666666666667`
2) Platform is X11 (Archlinux, KDE Plasma)

Also (and just to state the obvious), this only happens on my 4k monitor. When the DPI factor is 1, on my FHD one, everything is fine.

UHD monitor output:

width: 1920, height: 1200
width: 1920, height: 1200
surface_dimensions: [1920, 1200]
inner size: Some(LogicalSize { width: 1152.0, height: 720.0 })
outer_size: Some(LogicalSize { width: 1156.8, height: 739.8 })
dpi: 1.6666666666666667

FHD monitor output:
width: 1920, height: 1200
width: 1920, height: 1200
surface_dimensions: [1920, 1200]
inner size: Some(LogicalSize { width: 1920.0, height: 1200.0 })
outer_size: Some(LogicalSize { width: 1920.0, height: 1200.0 })
dpi: 1.0

Is there a problem in the calculation of logical and physical pixels, or am I doing something wrong? :)

**Comment by francesca64** [comment]
Your window size should be in logical pixels, whereas your surface should be in physical pixels (assuming that's somewhat analogous to an OpenGL viewport; I haven't yet had a chance to learn about Vulkan).

Do you still have the problem without using Vulkano, i.e. with a winit example program?

**Comment by DancingBard** [comment]
@francesca64 the problem initially came up, when I switched to `winit 0.16` (as elaborated) and was caused, because the `depthbuffe`r (to be more precise, the image representing the `depthbuffer`) was created with the wrong dimensions (logical pixels).

Vulkan(o) then failed during setup of the `framebuffers` as the `renderpass` (which describes where the output of the graphics piepeline will go [https://github.com/vulkano-rs/vulkano/blob/master/examples/src/bin/triangle.rs#L250]) had different dimensions than the `depthbuffer`... so I started investigating and found out the above behaviour.

Now all makes sense and with Winit alone I have no problems whatsoever. 
I am sorry for stealing your time but am thankful for the explanation. It totally makes sense!

In the end: mea culpa!
